include/boost/windows_sspi/stream.hpp

```diff
@@ -275,44 +257,22 @@ template <typename NextLayer> class stream : public stream_base {
 
   template <typename MutableBufferSequence>
   size_t read_some(const MutableBufferSequence& buffers, boost::system::error_code& ec) {
-    SECURITY_STATUS sc = SEC_E_OK;
-    while(m_sspi_impl->decrypt.decrypted_data.empty()) {
-      // TODO: This is duplicated in the async version. Move logic to
-      // sspi_impl some way. Possibly a decryption class for keep
-      // track of state and buffers.
-      std::vector<char> input_buffer;
-      if (m_sspi_impl->decrypt.encrypted_data.empty() || sc == SEC_E_INCOMPLETE_MESSAGE) {
-        input_buffer.resize(0x10000);
-        std::size_t size_read = m_next_layer.read_some(net::buffer(input_buffer.data(), input_buffer.size()), ec);
-        if (ec && size_read == 0 && m_sspi_impl->decrypt.decrypted_data.empty()) {
-          return 0;
-        }
-        input_buffer.resize(size_read);
-      }
-      sc = m_sspi_impl->decrypt(input_buffer);
-      input_buffer.clear();
-      if (sc == SEC_E_INCOMPLETE_MESSAGE) {
-        continue;
-      }
-      if (sc == SEC_I_CONTEXT_EXPIRED) {
-        // TODO: Shutdown the TLS context gracefully (implement
-        // shutdown), then return EOF.
-        ec = net::error::eof;
-        return 0;
-      }
-      if (sc != SEC_E_OK) {
-        ec = error::make_error_code(sc);
-        return 0;
-      }
-      break;
+    while(m_sspi_impl->decrypt() == detail::sspi_decrypt::state::data_needed) {
+      std::array<char, 0x10000> input_buffer;
+      std::size_t size_read = m_next_layer.read_some(net::buffer(input_buffer.data(), input_buffer.size()), ec);
```

Comment:
read_some used here whereas the async version used async_read[!_some] - I believe the sync version is correct.